<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="A function to convert the output of process_plink() into a big.matrix (such as could be passed to biglasso::biglasso())"><title>A function to convert the output of process_plink() into a big.matrix (such as could be passed to biglasso::biglasso()) — get_bm_from_plink • plmmr</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.11/clipboard.min.js" integrity="sha512-7O5pXpc0oCRrxk8RUfDYFgn0nO1t+jLuIOQdOMRp4APB7uZ4vSjspzp5y6YDtDs4VzUSTbWzBFZ/LKJhnyFOKw==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="A function to convert the output of process_plink() into a big.matrix (such as could be passed to biglasso::biglasso()) — get_bm_from_plink"><meta property="og:description" content="A function to convert the output of process_plink() into a big.matrix (such as could be passed to biglasso::biglasso())"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary" data-bs-theme="dark"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">plmmr</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">3.0.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href="../index.html">
    <span class="fa fa-home"></span>
     
  </a>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/getting-started.html">Getting started</a>
    <a class="dropdown-item" href="../articles/notation.html">Notes on notation</a>
    <a class="dropdown-item" href="../articles/filebacking.html">PLINK files/file-backed matrices</a>
    <a class="dropdown-item" href="../articles/deconfounding.html">In progress: PLMMs &amp; deconfounding</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">News</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/pbreheny/plmmr">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>A function to convert the output of <code>process_plink()</code> into a <code>big.matrix</code> (such as could be passed to <code>biglasso::biglasso()</code>)</h1>
      
      <div class="d-none name"><code>get_bm_from_plink.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>A function to convert the output of <code><a href="process_plink.html">process_plink()</a></code> into a <code>big.matrix</code> (such as could be passed to <code><a href="https://rdrr.io/pkg/biglasso/man/biglasso.html" class="external-link">biglasso::biglasso()</a></code>)</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">get_bm_from_plink</span><span class="op">(</span><span class="va">path</span>, standardize <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>path</dt>
<dd><p>The file path to the RDS object containing the processed data. Do not add the '.rds' extension to the path.</p></dd>


<dt>standardize</dt>
<dd><p>Logical: should the data be returned in a column-standardized form? Defaults to TRUE. Please don't change this unless you are super confident about what you are doing...</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>A list with elements:</p><ul><li><p>'X': a <code>big.matrix</code> object with the design matrix of the data. If standardize = TRUE, the data corresponding to <code>std_X</code> is returned.
Otherwise, the data corresponding to <code>subset_X</code> is returned</p></li>
<li><p>'map': data frame with <code>.bim</code> file information</p></li>
<li><p>'fam': data frame with <code>.fam</code> file information</p></li>
<li><p>'ns': A vector indicating the which columns of X contain nonsingular features (i.e., features with variance != 0.</p></li>
<li><p>'center': If standardize = TRUE, a vector of values for centering each column in X is returned</p></li>
<li><p>'scale': If standardize = TRUE, a vector of values for scaling each column in X is returned</p></li>
</ul></div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>This function is a wrapper combining <code><a href="get_data.html">get_data()</a></code> and <code><a href="fbm2bm.html">fbm2bm()</a></code></p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># \donttest{</span></span></span>
<span class="r-in"><span><span class="fu"><a href="process_plink.html">process_plink</a></span><span class="op">(</span>data_dir <span class="op">=</span> <span class="fu"><a href="get_example_data.html">get_example_data</a></span><span class="op">(</span>parent <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  prefix <span class="op">=</span> <span class="st">"penncath_lite"</span>,</span></span>
<span class="r-in"><span>  gz <span class="op">=</span> <span class="cn">TRUE</span>,</span></span>
<span class="r-in"><span>  outfile <span class="op">=</span> <span class="st">"process_penncath"</span>,</span></span>
<span class="r-in"><span>  <span class="co"># overwrite = TRUE, # uncomment if needed </span></span></span>
<span class="r-in"><span>  impute_method <span class="op">=</span> <span class="st">"mode"</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Logging to process_penncath.log</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Preprocessing penncath_lite data:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Creating  penncath_lite .rds</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Unzipping .gz files - this could take a second</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> There are  1401  observations and  4367  genomic features in the specified data files, representing chromosomes  1  -  22</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> There are a total of  3514 SNPs with missing values</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Of these,  13  are missing in at least 50% of the samples</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Will prune out  468  samples/observations with missing phenotype data.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Imputing the missing (genotype) values using  mode  method</span>
<span class="r-err co"><span class="r-pr">#&gt;</span> <span class="error">Error:</span> Two levels of parallelism are used. See `?assert_cores`.</span>
<span class="r-in"><span>  </span></span>
<span class="r-in"><span>  <span class="va">my_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="fu"><a href="get_example_data.html">get_example_data</a></span><span class="op">(</span>parent <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>, <span class="st">"/penncath_lite"</span><span class="op">)</span></span></span>
<span class="r-in"><span>  <span class="va">bm_data</span> <span class="op">&lt;-</span> <span class="fu">get_bm_from_plink</span><span class="op">(</span><span class="va">my_path</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Note: The design matrix is being returned as a file-backed matrix (FBM) -- see bigstatsr::FBM() for details.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Reminder: the X that is returned here is column-standardized.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> A copy of the original data is available via the 'genotypes' matrix in the .rds object</span>
<span class="r-err co"><span class="r-pr">#&gt;</span> <span class="error">Error in if (obj$id_var == "IID") {    row_names &lt;- as.character(obj$fam$sample.ID[obj$complete_phen])}:</span> argument is of length zero</span>
<span class="r-in"><span><span class="co"># }</span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Anna C. Reisetter, Patrick J. Breheny, Tabitha K. Peter, Yujing Lu.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

    </footer></div>

  

  

  </body></html>

