<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="Preprocess PLINK files using the bigsnpr package"><title>Preprocess PLINK files using the bigsnpr package — process_plink • plmmr</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.11/clipboard.min.js" integrity="sha512-7O5pXpc0oCRrxk8RUfDYFgn0nO1t+jLuIOQdOMRp4APB7uZ4vSjspzp5y6YDtDs4VzUSTbWzBFZ/LKJhnyFOKw==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Preprocess PLINK files using the bigsnpr package — process_plink"><meta property="og:description" content="Preprocess PLINK files using the bigsnpr package"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary" data-bs-theme="dark"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">plmmr</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">3.0.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href="../index.html">
    <span class="fa fa-home"></span>
     
  </a>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/getting-started.html">Getting started</a>
    <a class="dropdown-item" href="../articles/notation.html">Notes on notation</a>
    <a class="dropdown-item" href="../articles/filebacking.html">PLINK files/file-backed matrices</a>
    <a class="dropdown-item" href="../articles/deconfounding.html">In progress: PLMMs &amp; deconfounding</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">News</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/pbreheny/plmmr">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Preprocess PLINK files using the <code>bigsnpr</code> package</h1>
      
      <div class="d-none name"><code>process_plink.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Preprocess PLINK files using the <code>bigsnpr</code> package</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">process_plink</span><span class="op">(</span></span>
<span>  <span class="va">data_dir</span>,</span>
<span>  <span class="va">prefix</span>,</span>
<span>  rds_dir <span class="op">=</span> <span class="va">data_dir</span>,</span>
<span>  impute <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  impute_method <span class="op">=</span> <span class="st">"mode"</span>,</span>
<span>  na_phenotype_vals <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">9</span><span class="op">)</span>,</span>
<span>  id_var <span class="op">=</span> <span class="st">"IID"</span>,</span>
<span>  handle_missing_phen <span class="op">=</span> <span class="st">"prune"</span>,</span>
<span>  add_phen <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  pheno_id <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  pheno_name <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  quiet <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  gz <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  <span class="va">outfile</span>,</span>
<span>  overwrite <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  add_predictor_fam <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  add_predictor_ext <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  keep_bigSNP <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>data_dir</dt>
<dd><p>The path to the bed/bim/fam data files, <em>without</em> a trailing "/" (e.g., use <code>data_dir = '~/my_dir'</code>, <strong>not</strong> <code>data_dir = '~/my_dir/'</code>)</p></dd>


<dt>prefix</dt>
<dd><p>The prefix (as a character string) of the bed/fam data files (e.g., <code>prefix = 'mydata'</code>)</p></dd>


<dt>rds_dir</dt>
<dd><p>The path to the directory in which you want to create the new '.rds' and '.bk' files. Defaults to <code>data_dir</code></p></dd>


<dt>impute</dt>
<dd><p>Logical: should data be imputed? Default to TRUE.</p></dd>


<dt>impute_method</dt>
<dd><p>If 'impute' = TRUE, this argument will specify the kind of imputation desired. Options are:
* mode (default): Imputes the most frequent call. See <code><a href="https://privefl.github.io/bigsnpr/reference/snp_fastImputeSimple.html" class="external-link">bigsnpr::snp_fastImputeSimple()</a></code> for details.
* random: Imputes sampling according to allele frequencies.
* mean0: Imputes the rounded mean.
* mean2: Imputes the mean rounded to 2 decimal places.
* xgboost: Imputes using an algorithm based on local XGBoost models. See <code><a href="https://privefl.github.io/bigsnpr/reference/snp_fastImpute.html" class="external-link">bigsnpr::snp_fastImpute()</a></code> for details. Note: this can take several minutes, even for a relatively small data set.</p></dd>


<dt>na_phenotype_vals</dt>
<dd><p>A vector of numeric values used to code NA values in the phenotype/outcome (this is the 'affection' column in a <code>bigSNP</code> object, or the last column of a <code>.fam</code> file). Defaults to -9 (matching PLINK conventions).</p></dd>


<dt>id_var</dt>
<dd><p>String specifying which column of the PLINK <code>.fam</code> file has the unique sample identifiers. Options are "IID" (default) and "FID".</p></dd>


<dt>handle_missing_phen</dt>
<dd><p>A string indicating how missing phenotypes should be handled:
* "prune" (default): observations with missing phenotype are removed
* "median": impute missing phenotypes using the median (warning: this is overly simplistic in many cases).
* "mean": impute missing phenotypes using the mean (warning: this is overly simplistic in many cases).
Note: for data coming from PLINK, no missing values of the phenotype are allowed. You have to (1) supply phenotype from an external file,
(2) prune missing values, or (3) impute missing values.</p></dd>


<dt>add_phen</dt>
<dd><p>Optional: A <strong>data frame</strong> with at least two columns: and ID column and a phenotype column</p></dd>


<dt>pheno_id</dt>
<dd><p>Optional: A string specifying the name of the ID column in <code>pheno</code>. MUST be specified if <code>add_phen</code> is specified.</p></dd>


<dt>pheno_name</dt>
<dd><p>Optional: A string specifying the name of the phenotype column in <code>pheno</code>.  MUST be specified if <code>add_phen</code> is specified. This column will be used as the default <code>y</code> argument to 'plmm()'.</p></dd>


<dt>quiet</dt>
<dd><p>Logical: should messages to be printed to the console be silenced? Defaults to FALSE</p></dd>


<dt>gz</dt>
<dd><p>Logical: are the bed/bim/fam files g-zipped? Defaults to FALSE. NOTE: if TRUE, process_plink will unzip your zipped files.</p></dd>


<dt>outfile</dt>
<dd><p>Optional: the name (character string) of the prefix of the logfile to be written. Defaults to 'process_plink', i.e. you will get 'process_plink.log' as the outfile.</p></dd>


<dt>overwrite</dt>
<dd><p>Logical: if existing <code>.bk</code>/<code>.rds</code> files exist for the specified directory/prefix, should these be overwritten? Defaults to FALSE. Set to TRUE if you want to change the imputation method you're using, etc.</p></dd>


<dt>add_predictor_fam</dt>
<dd><p>Optional: if you want to include "sex" (the 5th column of <code>.fam</code> file) in the analysis, specify 'sex' here.</p></dd>


<dt>add_predictor_ext</dt>
<dd><p>Optional: add additional covariates/predictors/features from an external file (i.e., not a PLINK file).
This argument takes one of two kinds of arguments:
- a <strong>named</strong> numeric vector, where the names align with the sample IDs in the PLINK files.
The names will be used to subset and align this external covariate with the supplied PLINK data.</p>
<p></p><div class="sourceCode"><pre><code><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a>                         <span class="sc">-</span> a numeric matrix whose row names align with the sample IDs <span class="cf">in</span> the PLINK files. </span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a>                      The names will be used to subset and align this external covariate with the supplied PLINK data.</span></code></pre><p></p></div></dd>


<dt>keep_bigSNP</dt>
<dd><p>Logical: should the intermediate steps of data processing be saved as a <code>bigSNP</code> object? Defaults to FALSE.</p></dd>


<dt>...</dt>
<dd><p>Optional: additional arguments to <code><a href="https://privefl.github.io/bigsnpr/reference/snp_fastImpute.html" class="external-link">bigsnpr::snp_fastImpute()</a></code> (relevant only if impute_method = "xgboost")</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>Nothing is returned by this function, but (at least) two files are created in
the location specified by <code>rds_dir</code>:</p><ul><li><p>'std_prefix.rds': This is the <code><a href="https://privefl.github.io/bigsnpr/reference/bigSNP-class.html" class="external-link">bigsnpr::bigSNP</a></code> object
that holds the PLINK data along with meta-data. See details for explanation of what
is included in this meta-data</p></li>
<li><p>'std_prefix.bk': Created by the call to <code><a href="standardize_fbm.html">standardize_fbm()</a></code>, this is the
backingfile that stores the numeric data of the standardized design matrix <code>std_X</code></p></li>
</ul><p>Intermediate files 'prefix.rds' and 'prefix.bk' are also created along the way;
if <code>keep_bigSNP = TRUE</code>, these are not deleted at the end of the <code>process_plink()</code></p>


<p>procedure. Note that these files could potentially be quite large - that's the
main reason that keeping these files is not the default setting.</p>


<p>Note that <code>process_plink()</code> need only be run once for a given set of PLINK
files; in subsequent data analysis/scripts, <code><a href="get_data.html">get_data()</a></code> will access the '.rds' file.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>The '.rds' object created by this function has the following elements:</p><ul><li><p>std_X: The file-backed design matrix, as an <code>FBM</code> object (see <code>bigstatsr</code> package for more info)</p></li>
<li><p>fam: Data frame equivalent of PLINK '.bim' file</p></li>
<li><p>map: Data frame equivalent of PLINK '.fam' file</p></li>
<li><p>colnames: Character vector of column names for the original data (includes constant features, i.e. monomorphic SNPs)</p></li>
<li><p>rownames: Character vector of row names for the original data.</p></li>
<li><p>n: Number of rows (samples) in the original data</p></li>
<li><p>p: Number of columns (features, SNPs, markers, ...) in the original data</p></li>
<li><p>ns: Numeric vector of indices marking the non-singular columns of the original data</p></li>
<li><p>std_X_center: Numeric vector of values used to center the non-singular columns of the data</p></li>
<li><p>std_X_scale: Numeric vector of values used to scale the non-singular columns of the data</p></li>
<li><p>std_X_colnames: Character vector of column names for the standardized data</p></li>
<li><p>std_X_rownames: Character vector of row names for the standardized data.</p></li>
<li><p>complete_phen: Numeric vector of indices marking the samples with a non-missing phenotype. Ony applicable if <code>handle_missing_phen = 'prune'</code></p></li>
<li><p>id_var: String specifying which ID column in the '.fam' file had the unique sample ID: 'FID' (1st column) or 'IID' (2nd column)</p></li>
</ul></div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># \donttest{</span></span></span>
<span class="r-in"><span><span class="va">my_dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu">process_plink</span><span class="op">(</span>data_dir <span class="op">=</span> <span class="fu"><a href="get_example_data.html">get_example_data</a></span><span class="op">(</span>parent <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  rds_dir <span class="op">=</span> <span class="va">my_dir</span>,</span></span>
<span class="r-in"><span>  prefix <span class="op">=</span> <span class="st">"penncath_lite"</span>,</span></span>
<span class="r-in"><span>  gz <span class="op">=</span> <span class="cn">TRUE</span>,</span></span>
<span class="r-in"><span>  outfile <span class="op">=</span> <span class="st">"process_penncath"</span>,</span></span>
<span class="r-in"><span>  overwrite <span class="op">=</span> <span class="cn">TRUE</span>,</span></span>
<span class="r-in"><span>  impute_method <span class="op">=</span> <span class="st">"mode"</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Logging to process_penncath.log</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Preprocessing penncath_lite data:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Overwriting existing files:  penncath_lite .bk/.rds</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Creating  penncath_lite .rds</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> The file  /home/runner/work/_temp/Library/plmmr/extdata/penncath_lite.bed  already exists. Setting gz = FALSE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> There are  1401  observations and  4367  genomic features in the specified data files, representing chromosomes  1  -  22</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> There are a total of  3514 SNPs with missing values</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Of these,  13  are missing in at least 50% of the samples</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Will prune out  468  samples/observations with missing phenotype data.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Imputing the missing (genotype) values using  mode  method</span>
<span class="r-err co"><span class="r-pr">#&gt;</span> <span class="error">Error:</span> Two levels of parallelism are used. See `?assert_cores`.</span>
<span class="r-in"><span><span class="co"># }</span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Anna C. Reisetter, Patrick J. Breheny, Tabitha K. Peter, Yujing Lu.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

    </footer></div>

  

  

  </body></html>

