<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="plmmr">
<title>Analyze high-dimensional PLINK data with filebacking • plmmr</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.11/clipboard.min.js" integrity="sha512-7O5pXpc0oCRrxk8RUfDYFgn0nO1t+jLuIOQdOMRp4APB7uZ4vSjspzp5y6YDtDs4VzUSTbWzBFZ/LKJhnyFOKw==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="Analyze high-dimensional PLINK data with filebacking">
<meta property="og:description" content="plmmr">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary" data-bs-theme="dark"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">plmmr</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">3.0.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../index.html">
    <span class="fa fa-home"></span>
     
  </a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/getting-started.html">Getting started</a>
    <a class="dropdown-item" href="../articles/notation.html">Notes on notation</a>
    <a class="dropdown-item" href="../articles/filebacking.html">PLINK files/file-backed matrices</a>
    <a class="dropdown-item" href="../articles/deconfounding.html">In progress: PLMMs &amp; deconfounding</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">News</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/pbreheny/plmmr">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Analyze high-dimensional PLINK data with filebacking</h1>
            
      
      
      <div class="d-none name"><code>filebacking.Rmd</code></div>
    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">plmmr</span><span class="op">)</span></span>
<span><span class="co"># MUST have biglasso version &gt;= 1.5.2.1 - uncomment &amp; run line below if needed</span></span>
<span><span class="co"># remotes::install_github(repo = "YaohuiZeng/biglasso") </span></span>
<span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://pbreheny.github.io/biglasso/index.html" class="external-link">biglasso</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://privefl.github.io/bigsnpr/" class="external-link">bigsnpr</a></span><span class="op">)</span></span></code></pre></div>
<p>In many applications of high dimensional data analysis, the dataset
is too large to read into R – the session will crash for lack of memory.
This is almost always an issue with analyzing genetic data from <a href="https://www.cog-genomics.org/plink/" class="external-link">PLINK</a> files. To analyze
such large datasets, <code>plmmr</code> is equipped to analyze data
using <em>filebacking</em> - a strategy that lets R ‘point’ to a file on
disk, rather than reading the file into the R session. Many other
packages use this technique - <a href="https://privefl.github.io/bigstatsr/" class="external-link">bigstatsr</a> and <a href="https://github.com/pbreheny/biglasso" class="external-link">biglasso</a> are two
examples of packages that use filebacking. The novelties of
<code>plmmr</code> are:</p>
<ol style="list-style-type: decimal">
<li><p><strong>Integration</strong>: <code>plmmr</code> combines the
functionality of several packages in order to do quality control, model
fitting/analysis, and data visualization all in one package.
<code>plmmr</code> will take you from PLINK files all the way to a list
of SNPs for downstream analysis.</p></li>
<li><p><strong>Accessibility</strong>: <code>plmmr</code> can be run
from an <code>R</code> session on a typical desktop/laptop machine. The
user does not need access to a supercomputer or experience with the
command line in order to fit models <code>plmmr</code>. If you don’t
know what the ‘command line’ means, don’t worry - <code>plmmr</code> was
designed with you in mind :)</p></li>
<li><p><strong>Handling correlation</strong>: <code>plmmr</code> uses a
transformation that addresses correlation among samples and uses this
correlation to improve predictions (via the best linear unbiased
predictor, or BLUP). This means that in <code><a href="../reference/plmm.html">plmm()</a></code>, there’s no
need to filter data down to a ‘maximum subset of unrelated
individuals.’</p></li>
</ol>
<p>The tutorial that follows will walk you through each step of the
process. We will use the <code>penncath_lite</code> data that ships with
our <code>plmmr</code> package – if you’ve installed this package, you
have this dataset as a set of PLINK files. These PLINK files represent
about 1400 participants and 4,000+ SNPs. This dataset is small enough
that we could read it into memory (4,000 is a small number of SNPs to
study), but I will analyze it using filebacked methods here for the sake
of example. The steps shown here would be the same for GWAS-sized
datasets (100,000 - 800,000 SNPs). More information about the
<code>penncath_lite</code> data is given in the <code>PLINK files</code>
article.</p>
<div class="section level2">
<h2 id="analyzing-filebacked-data-from-start-to-finish">Analyzing filebacked data from start to finish<a class="anchor" aria-label="anchor" href="#analyzing-filebacked-data-from-start-to-finish"></a>
</h2>
<div class="section level3">
<h3 id="step-1-preprocess-the-data">Step 1: preprocess the data<a class="anchor" aria-label="anchor" href="#step-1-preprocess-the-data"></a>
</h3>
<p>The most important step is where you begin – for GWAS data, we have
to tell <code>plmmr</code> how to combine information across all three
PLINK files (the <code>.bed</code>, <code>.bim</code>, and
<code>.fam</code> files). Here, we will create the files we want in a
temporary directory just for the sake of example. Users can specify the
folder of their choice for <code>rds_dir</code>, as shown below:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">temp_dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>,<span class="st">"/"</span><span class="op">)</span> <span class="co"># using a temporary directory</span></span>
<span><span class="fu"><a href="../reference/process_plink.html">process_plink</a></span><span class="op">(</span>data_dir <span class="op">=</span> <span class="fu"><a href="../reference/get_example_data.html">get_example_data</a></span><span class="op">(</span>parent <span class="op">=</span> <span class="cn">T</span><span class="op">)</span>, <span class="co"># reads data from inst/extdata</span></span>
<span>              rds_dir <span class="op">=</span> <span class="va">temp_dir</span>,</span>
<span>              prefix <span class="op">=</span> <span class="st">"penncath_lite"</span>,</span>
<span>              gz <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>              outfile <span class="op">=</span> <span class="st">"process_penncath"</span>,</span>
<span>              overwrite <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>              impute_method <span class="op">=</span> <span class="st">"mode"</span><span class="op">)</span></span></code></pre></div>
<p>You’ll see a lot of messages printed to the console here … the result
of all this is the creation of 2 files (these will show up in the folder
where the PLINK data is): <code>std_penncath_lite.rds</code> and
<code>std_penncath_lite.bk.</code></p>
<p>We can examine what’s in <code>std_penncath_lite.rds</code> using the
<code><a href="https://privefl.github.io/bigsnpr/reference/snp_attach.html" class="external-link">snp_attach()</a></code> function from the <code>bigsnpr</code>
package:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pen</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">temp_dir</span>, <span class="st">"std_penncath_lite.rds"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">pen</span><span class="op">)</span> <span class="co"># note: genotypes and std_X are *not* in memory</span></span>
<span><span class="co"># </span></span>
<span><span class="co"># we can check to see that our data have been standardized </span></span>
<span><span class="va">std_X</span> <span class="op">&lt;-</span> <span class="va">pen</span><span class="op">$</span><span class="va">std_X</span><span class="op">[</span>,<span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/colSums-methods.html" class="external-link">colMeans</a></span><span class="op">(</span><span class="va">std_X</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="op">)</span> <span class="co"># columns have mean zero...</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="va">std_X</span>, <span class="fl">2</span>, <span class="va">var</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="op">)</span> <span class="co"># ... &amp; variance 1</span></span></code></pre></div>
<div class="section level4">
<h4 id="what-if-my-data-arent-in-plink-format">What if my data aren’t in PLINK format?<a class="anchor" aria-label="anchor" href="#what-if-my-data-arent-in-plink-format"></a>
</h4>
</div>
</div>
</div>
<div class="section level2">
<h2 id="the-methodology-in-plmmr-was-motivated-by-applications-in-gwas-but-it-certainly-isnt-limited-to-that-context--suppose-you-have-a-high-dimemsional-data-set-such-as-the-colon2-data-this-is-like-the-colon-data-that-ships-with-the-biglasso-package-only-here-ive-inserted-some-na-values-just-for-didactic-purposes-">The methodology in <code>plmmr</code> was motivated by applications
in GWAS, but it certainly isn’t limited to that context. Suppose you
have a high dimemsional data set, such as the <code>colon2</code> data –
this is like the <code>colon</code> data that ships with the
<code>biglasso</code> package, only here I’ve inserted some NA values
just for didactic purposes.<a class="anchor" aria-label="anchor" href="#the-methodology-in-plmmr-was-motivated-by-applications-in-gwas-but-it-certainly-isnt-limited-to-that-context--suppose-you-have-a-high-dimemsional-data-set-such-as-the-colon2-data-this-is-like-the-colon-data-that-ships-with-the-biglasso-package-only-here-ive-inserted-some-na-values-just-for-didactic-purposes-"></a>
</h2>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/process_X.html">process_X</a></span><span class="op">(</span>file <span class="op">=</span> <span class="st">"colon2.txt"</span>,</span>
<span>          data_dir <span class="op">=</span> <span class="fu"><a href="../reference/get_example_data.html">get_example_data</a></span><span class="op">(</span>parent <span class="op">=</span> <span class="cn">T</span><span class="op">)</span>,</span>
<span>          rds_dir <span class="op">=</span> <span class="va">temp_dir</span>,</span>
<span>          ind.col <span class="op">=</span> <span class="fl">2</span><span class="op">:</span><span class="fl">2002</span><span class="op">)</span> <span class="co"># don't use the 1st column; it has IDs in it</span></span>
<span></span>
<span><span class="va">colon2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span>file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">temp_dir</span>, <span class="st">"std_colon2.rds"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">colon2</span><span class="op">)</span> <span class="co"># notice that the columns with NA values were dropped </span></span></code></pre></div>
<div class="section level3">
<h3 id="step-2-fit-a-model">Step 2: fit a model<a class="anchor" aria-label="anchor" href="#step-2-fit-a-model"></a>
</h3>
<p>Note: for data coming from <code><a href="../reference/process_plink.html">process_plink()</a></code>, the
<code>y</code> argument in <code><a href="../reference/plmm.html">plmm()</a></code> defaults to using the 6th
column from the ‘.fam’ file as the phenotype/outcome (this is the same
as what PLINK does).</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">my_fb_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">temp_dir</span>, <span class="st">"std_penncath_lite"</span><span class="op">)</span></span>
<span><span class="va">fb_fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plmm.html">plmm</a></span><span class="op">(</span>X <span class="op">=</span> <span class="va">my_fb_data</span>,</span>
<span>               returnX <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>               <span class="co"># this datset is small enough to fit in memory</span></span>
<span>               <span class="co"># by setting returnX = FALSE, I force plmm() to run on the filebacked data</span></span>
<span>               trace <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>The format is similar for data that came from <code>process_X</code>
– however, in the case of our <code>colon2</code> data, we need to
supply our own kinship matrix. The data are not genomic, so estimating a
genomic relatedness matrix doesn’t make sense. For our purposes, we will
make up a ‘dummy’ K matrix.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># pretend genomic data from 500 SNPs</span></span>
<span><span class="va">genomic_dat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fl">0</span><span class="op">:</span><span class="fl">2</span>, <span class="fl">62</span><span class="op">*</span><span class="fl">500</span>, replace <span class="op">=</span> <span class="cn">T</span><span class="op">)</span>, nrow <span class="op">=</span> <span class="fl">62</span>, ncol <span class="op">=</span> <span class="fl">500</span><span class="op">)</span></span>
<span><span class="va">K</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/relatedness_mat.html">relatedness_mat</a></span><span class="op">(</span><span class="va">genomic_dat</span><span class="op">)</span></span>
<span></span>
<span><span class="va">colon2_fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plmm.html">plmm</a></span><span class="op">(</span>X <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">temp_dir</span>, <span class="st">"std_colon2"</span><span class="op">)</span>,</span>
<span>                   y <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">rnorm</a></span><span class="op">(</span><span class="fl">62</span><span class="op">)</span>,</span>
<span>                   K <span class="op">=</span> <span class="va">K</span>,</span>
<span>                   returnX <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                   trace <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span></code></pre></div>
<p>In most applications, we will want to fit a model using cross
validation</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cv_fb_fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/cv.plmm.html">cv.plmm</a></span><span class="op">(</span>X <span class="op">=</span> <span class="va">my_fb_data</span>,</span>
<span>                     type <span class="op">=</span> <span class="st">'blup'</span>,</span>
<span>                     returnX <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                     nfolds <span class="op">=</span> <span class="fl">3</span>, <span class="co"># for sake of testing, I made this small. </span></span>
<span>                     <span class="co"># In practice, 3 folds is usually not enough for robust </span></span>
<span>                     <span class="co"># cross validation - hence, the default is nfolds = 10</span></span>
<span>                     trace <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="step-3-plotsummarize-results">Step 3: Plot/summarize results<a class="anchor" aria-label="anchor" href="#step-3-plotsummarize-results"></a>
</h3>
<p>To summarize a single model (no CV), we can look at a plot of the
coefficient paths:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">fb_fit</span><span class="op">)</span></span></code></pre></div>
<p>We can also examine the number of selected variables at a particular
index of <span class="math inline">\(\lambda\)</span>:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">fb_fit</span>, idx <span class="op">=</span> <span class="fl">50</span><span class="op">)</span></span></code></pre></div>
<p>To summarize and plot a CV fit, we can look at a plot of cross
validation error:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">cv_fb_fit</span><span class="op">)</span></span></code></pre></div>
<p>To summarize and visualize the model at the chosen <span class="math inline">\(\lambda\)</span> parameter value, we can do
this:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">cv_fb_fit</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="comparing-candidate-models">Comparing candidate models<a class="anchor" aria-label="anchor" href="#comparing-candidate-models"></a>
</h3>
<p>The default penalty in <code>plmmr</code> is the minimax-concave
penalty (MCP)<a class="footnote-ref" tabindex="0" data-bs-toggle="popover" data-bs-content="&lt;p&gt;Zhang, Cun-Hui. “Nearly unbiased variable selection
under minimax concave penalty.” (2010): 894-942.&lt;/p&gt;"><sup>1</sup></a>. Suppose we want to compare our
MCP-penalized model results with a lasso-penalized model. We can fit a
lasso model as shown:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fb_lasso</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plmm.html">plmm</a></span><span class="op">(</span>X <span class="op">=</span> <span class="va">my_fb_data</span>,</span>
<span>                 penalty <span class="op">=</span> <span class="st">"lasso"</span>,</span>
<span>                 returnX <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                 trace <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">fb_lasso</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">fb_lasso</span>, idx <span class="op">=</span> <span class="fl">50</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co"># with CV </span></span>
<span><span class="va">cv_fb_lasso</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/cv.plmm.html">cv.plmm</a></span><span class="op">(</span>X <span class="op">=</span> <span class="va">my_fb_data</span>,</span>
<span>                       penalty <span class="op">=</span> <span class="st">"lasso"</span>,</span>
<span>                     type <span class="op">=</span> <span class="st">'blup'</span>,</span>
<span>                     nfolds <span class="op">=</span> <span class="fl">3</span>, </span>
<span>                     returnX <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                     trace <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">cv_fb_lasso</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">cv_fb_lasso</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="to-add-predictors-that-are-not-genomic">To add predictors that are not genomic<a class="anchor" aria-label="anchor" href="#to-add-predictors-that-are-not-genomic"></a>
</h3>
<p>In many biological applications, we want to include some
features/covariates/predictors in addition to the genomic information,
like clinical or demographic information. Let’s suppose that in the
<code>penncath</code> data analysis, we want to include ‘sex’ and ‘age’
as unpenalized covariates. We can do this by supplying an additional
argument to <code><a href="../reference/process_plink.html">process_plink()</a></code>:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pen_clinic</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="fu"><a href="../reference/get_example_data.html">get_example_data</a></span><span class="op">(</span>parent <span class="op">=</span> <span class="cn">T</span><span class="op">)</span>, <span class="st">"/penncath_clinical.csv"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">extdata</span> <span class="op">&lt;-</span> <span class="va">pen_clinic</span><span class="op">[</span>,<span class="fl">3</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">extdata</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">pen_clinic</span><span class="op">$</span><span class="va">FamID</span> <span class="co"># This is important! </span></span>
<span></span>
<span><span class="co"># create a new temporary directory</span></span>
<span><span class="va">temp_dir2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="va">LETTERS</span>, <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/process_plink.html">process_plink</a></span><span class="op">(</span>data_dir <span class="op">=</span> <span class="fu"><a href="../reference/get_example_data.html">get_example_data</a></span><span class="op">(</span>parent <span class="op">=</span> <span class="cn">T</span><span class="op">)</span>,</span>
<span>              rds_dir <span class="op">=</span> <span class="va">temp_dir2</span>, <span class="co"># using a temporary directory</span></span>
<span>              prefix <span class="op">=</span> <span class="st">"penncath_lite"</span>,</span>
<span>              id_var <span class="op">=</span> <span class="st">"FID"</span>, <span class="co"># this is KEY!</span></span>
<span>              outfile <span class="op">=</span> <span class="st">"process_penncath"</span>,</span>
<span>              impute_method <span class="op">=</span> <span class="st">"mode"</span>,</span>
<span>              add_predictor_ext <span class="op">=</span> <span class="va">extdata</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># check this out: </span></span>
<span><span class="va">pen2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">temp_dir2</span>, <span class="st">"/std_penncath_lite.rds"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">pen2</span><span class="op">$</span><span class="va">std_X_colnames</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span> <span class="co"># std_X includes our non-genomic covariates </span></span></code></pre></div>
<p>We usually want these kinds of variables to always be in the model
(i.e., they are not penalized). To make sure ‘sex’ and ‘age’ are always
included in the chosen model, let’s set <code>penalty.factor</code> to
have 0 values corresponding to these predictors.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dat_plus_newvars</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">temp_dir2</span>, <span class="st">"/std_penncath_lite"</span><span class="op">)</span></span>
<span><span class="va">fit_plus_newvars</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plmm.html">plmm</a></span><span class="op">(</span>X <span class="op">=</span> <span class="va">dat_plus_newvars</span>,</span>
<span>               penalty.factor <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0</span>, <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">pen2</span><span class="op">$</span><span class="va">std_X</span><span class="op">)</span> <span class="op">-</span> <span class="fl">2</span><span class="op">)</span><span class="op">)</span>,</span>
<span>               returnX <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>               trace <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>We can check out how these additional covariates impact model
fit:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">fit_plus_newvars</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">fb_fit</span>, idx <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> </span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">fit_plus_newvars</span>, idx <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="co"># at highest penalty value, sex and age are still in the model</span></span>
<span></span>
<span><span class="co"># look at the estimated coefficients </span></span>
<span><span class="va">fb_fit</span><span class="op">$</span><span class="va">beta_vals</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span>
<span><span class="va">fit_plus_newvars</span><span class="op">$</span><span class="va">beta_vals</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span></code></pre></div>
</div>
</div>

  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Anna C. Reisetter, Patrick J. Breheny, Tabitha K. Peter, Yujing Lu.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
