<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Getting started with plmmr • plmmr</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="Getting started with plmmr">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">plmmr</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">3.2.0.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../index.html"><span class="fa fa-home"></span></a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/getting-started.html">Start here - getting to know plmmr</a></li>
    <li><a class="dropdown-item" href="../articles/matrix_data.html">If your data is a matrix or data frame</a></li>
    <li><a class="dropdown-item" href="../articles/plink_files.html">If your data is in PLINK files</a></li>
    <li><a class="dropdown-item" href="../articles/delim_files.html">If your data is in (a) delimited file(s)</a></li>
    <li><a class="dropdown-item" href="../articles/notation.html">Notes on notation</a></li>
    <li><a class="dropdown-item" href="../articles/deconfounding.html">In progress: PLMMs &amp; deconfounding</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">News</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/pbreheny/plmmr"><span class="fa fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Getting started with plmmr</h1>
                        <h4 data-toc-skip class="author">Tabitha
Peter</h4>
            
      

      <div class="d-none name"><code>getting-started.Rmd</code></div>
    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">plmmr</span><span class="op">)</span></span>
<span><span class="co">#&gt; Loading required package: bigalgebra</span></span>
<span><span class="co">#&gt; Loading required package: bigmemory</span></span></code></pre></div>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p><code>plmmr</code> (I say it as “PLUM-are”) is a package for fitting
<strong>P</strong>enalized <strong>L</strong>inear
<strong>M</strong>ixed <strong>M</strong>odels in <strong>R</strong>.
This package was created for the purpose of fitting penalized regression
models to high dimensional data in which the observations are correlated
in some complex way. For instance, this kind of data arises often in the
context of genetics (<em>e.g.</em>, GWAS dealing with population
structure/family structure).</p>
<p>For more information on the theoretical context of penalized
regression, check out <a href="https://myweb.uiowa.edu/pbreheny/7240/s23/notes.html" class="external-link">these online
course notes</a> from Prof. Patrick Breheny (the PI of our lab). The
section on ‘ridge regression’ may be a good place to start.</p>
<p>The novelties of <code>plmmr</code> are:</p>
<ol style="list-style-type: decimal">
<li><p><strong>Integration</strong>: <code>plmmr</code> combines the
functionality of several packages in order to do quality control, model
fitting/analysis, and data visualization all in one package.
<code>plmmr</code> will take you from PLINK files all the way to a list
of SNPs for downstream analysis.</p></li>
<li><p><strong>Accessibility</strong>: <code>plmmr</code> can be run
from an <code>R</code> session on a typical desktop/laptop machine. The
user does not need access to a supercomputer or experience with the
command line in order to fit models <code>plmmr</code>.</p></li>
<li><p><strong>Handling correlation</strong>: <code>plmmr</code> uses a
transformation that addresses correlation among samples and uses this
correlation to improve predictions (via the best linear unbiased
predictor, or BLUP). This means that in <code><a href="../reference/plmm.html">plmm()</a></code>, there’s no
need to filter data down to a ‘maximum subset of unrelated
individuals.’</p></li>
</ol>
</div>
<div class="section level2">
<h2 id="computational-capability">Computational capability<a class="anchor" aria-label="anchor" href="#computational-capability"></a>
</h2>
<div class="section level3">
<h3 id="file-backing">File-backing<a class="anchor" aria-label="anchor" href="#file-backing"></a>
</h3>
<p>In many applications of high dimensional data analysis, the dataset
is too large to read into R – the session will crash for lack of memory.
This is almost always an issue with analyzing genetic data from <a href="https://www.cog-genomics.org/plink/" class="external-link">PLINK</a> files. To analyze
such large datasets, <code>plmmr</code> is equipped to analyze data
using <em>filebacking</em> - a strategy that lets R ‘point’ to a file on
disk, rather than reading the file into the R session. Many other
packages use this technique - <a href="https://privefl.github.io/bigstatsr/" class="external-link">bigstatsr</a> and <a href="https://github.com/pbreheny/biglasso" class="external-link">biglasso</a> are two
examples of packages that use filebacking.</p>
</div>
<div class="section level3">
<h3 id="numeric-outcomes-only">Numeric outcomes only<a class="anchor" aria-label="anchor" href="#numeric-outcomes-only"></a>
</h3>
<p>At this time, the package is designed for linear regression only –
that is, we are considering only continuous (numeric) outcomes. We
maintain that treating binary or categorical outcomes as numeric values
is appropriate in some contexts, as described by Hastie et al. in the <a href="https://hastie.su.domains/Papers/ESLII.pdf" class="external-link">Elements of
Statistical Learning</a>. In the future, we would like to extend this
package to handle dichotomous outcomes via logistic regression; the
theoretical work underlying this is an open problem.</p>
</div>
<div class="section level3">
<h3 id="types-of-penalization">3 types of penalization<a class="anchor" aria-label="anchor" href="#types-of-penalization"></a>
</h3>
<p>Since we are focused on penalized regression in this package,
<code>plmmr</code> offers 3 choices of penalty: the minimax concave
(MCP), the smoothly clipped absolute deviation (SCAD), and the least
absolute shrinkage and selection operator (LASSO). Much of the work in
this package is built on the concepts/techniques provided in the
<code>ncvreg</code> <a href="https://github.com/pbreheny/ncvreg" class="external-link">package</a> by <a href="https://myweb.uiowa.edu/pbreheny/" class="external-link">Patrick Breheny</a>. The
filebacked computation relies on the <code>biglasso</code> <a href="https://github.com/pbreheny/biglasso" class="external-link">package</a> by <a href="https://scholar.google.com/citations?user=jpEmf04AAAAJ&amp;hl=en" class="external-link">Yaohui
Zeng</a> &amp; colleagues. For processing PLINK files, we use methods
from the <code>bigsnpr</code> <a href="https://privefl.github.io/bigsnpr/" class="external-link">package</a> by <a href="https://privefl.github.io/" class="external-link">Florian Privé</a>.</p>
</div>
<div class="section level3">
<h3 id="data-size-and-dimensionality">Data size and dimensionality<a class="anchor" aria-label="anchor" href="#data-size-and-dimensionality"></a>
</h3>
<p>We distinguish between the data attributes ‘big’ and ‘high
dimensional.’ ‘Big’ describes the amount of space data takes up on a
computer, while ‘high dimensional’ describes a context where the ratio
of features (also called ‘variables’ or ‘predictors’) to observations
(e.g., samples) is high. For instance, data with 100 samples and 100
variables is high dimensional, but not big. By contrast, data with 10
million observations and 100 variables is big, but not high
dimensional.</p>
<p><code>plmmr</code> is optimized for data that are high dimensional –
the methods we are using to estimate relatedness among observations
perform best when there are a high number of features relative to the
number of observations.</p>
<p><code>plmmr</code> is also designed to accommodate data that is too
large to analyze in-memory. We accommodate such data through
file-backing (as described above). This approach works well for data
with files up to about 40 Gb in size – our current methods are not
optimized for data larger that 40 Gb. In practice, this means that
<code>plmmr</code> is equipped to analyze GWAS data, but not
biobank-sized data.</p>
</div>
</div>
<div class="section level2">
<h2 id="data-input-types">Data input types<a class="anchor" aria-label="anchor" href="#data-input-types"></a>
</h2>
<p><code>plmmr</code> currently works with three types of data
input:</p>
<ol style="list-style-type: decimal">
<li><p>Data that is stored in-memory as a matrix or data frame</p></li>
<li><p>Data that is stored in PLINK files</p></li>
<li><p>Data that is stored in delimited files</p></li>
</ol>
<div class="section level3">
<h3 id="example-data-sets">Example data sets<a class="anchor" aria-label="anchor" href="#example-data-sets"></a>
</h3>
<p><code>plmmr</code> currently includes three example data sets, one
for each type of data input.</p>
<ul>
<li><p>The <code>admix</code> data is our example of matrix input data.
<code>admix</code> is a small data set (197 observations, 100 SNPs) that
describes individuals of different ancestry groups. The outcome of
<code>admix</code> is simulated to include population structure effects
(<em>i.e.</em> race/ethnicity have an impact on the SNP associations).
This data set is available as whenever <code><a href="https://rdrr.io/r/base/library.html" class="external-link">library(plmmr)</a></code> is
called. An example analysis with the <code>admix</code> data is
available in
<code>vignette('matrix_data', package = "plmmr")</code>.</p></li>
<li><p>The <code>penncath_lite</code> data is our example of PLINK input
data. <code>penncath_lite</code> (data on coronary artery disease from
the <a href="https://pubmed.ncbi.nlm.nih.gov/21239051/" class="external-link">PennCath
study</a>) is a mid-sized, high dimensional data set (1401 observations,
4217 SNPs) with several health outcomes as well as age and sex
information. This data set is a subset of a much larger data set (the
original data has over 800K SNPs). For for information on this data set,
refer to the <a href="https://pubmed.ncbi.nlm.nih.gov/21239051/" class="external-link">original
publication</a>. An example analysis with the <code>penncath_lite</code>
data is available in
<code>vignette('plink_files', package = "plmmr")</code>.</p></li>
<li><p>The <code>colon2</code> data is our example of delimited-file
input data. <code>colon2</code> is a variation of the <code>colon</code>
data included in the <a href="LINK"><code>biglasso</code> package</a>.
<code>colon2</code> has 62 observations and 2,001 features representing
a study of colon disease. 2000 features are original to the data, and
the ‘sex’ feature is simulated. An example analysis with the
<code>colon2</code> data is available in
<code>vignette('delim_files', package = "plmmr")</code>.</p></li>
</ul>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Tabitha K. Peter, Anna C. Reisetter, Patrick J. Breheny, Yujing Lu.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
