<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Notes on analyzing data from PLINK files • plmmr</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="Notes on analyzing data from PLINK files">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">plmmr</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">3.1.1.3</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../index.html"><span class="fa fa-home"></span></a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/getting-started.html">Getting started</a></li>
    <li><a class="dropdown-item" href="../articles/process_plink_files.html">If your data is from PLINK...</a></li>
    <li><a class="dropdown-item" href="../articles/notation.html">Notes on notation</a></li>
    <li><a class="dropdown-item" href="../articles/filebacking.html">File-backed data (how to work with big files)</a></li>
    <li><a class="dropdown-item" href="../articles/deconfounding.html">In progress: PLMMs &amp; deconfounding</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">News</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/pbreheny/plmmr"><span class="fa fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Notes on analyzing data from PLINK files</h1>
            
      

      <div class="d-none name"><code>processing_plink_files.Rmd</code></div>
    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">plmmr</span><span class="op">)</span></span>
<span><span class="co">#&gt; Loading required package: bigalgebra</span></span>
<span><span class="co">#&gt; Loading required package: bigmemory</span></span></code></pre></div>
<p>A widely-used format for storing data from genome-wide association
studies (GWAS) is the , based on the PLINK software. The
<code>plmmr</code> package is equipped to analyze data from one of the
<a href="https://www.cog-genomics.org/plink/1.9/formats" class="external-link">PLINK file
formats</a>: the triplet of ‘.bed’, ‘.bed’, and ‘.fam’ files. If you
have data in this format, keep reading – if you want to know more about
what each of these files contains, see <a href="https://pbreheny.github.io/adv-gwas-tutorial/quality_control.html" class="external-link">this
other tutorial</a>.</p>
<p>When analyzing PLINK data, the approach <code>plmmr</code> takes
involves these steps:</p>
<ul>
<li><p>Reading in all 3 PLINK data files and combining their information
into a <a href="https://privefl.github.io/bigsnpr/reference/bigSNP-class.html" class="external-link">bigSNP
object</a>.</p></li>
<li><p>Integrate in the external phenotype information, if supplied.
<strong>Note: Any samples in the PLINK data that do not have a phenotype
value in the specified additional phenotype file will be removed from
the analysis.</strong> If no external phenotype value is supplied, the
outcome defaults to the 6th column of the ‘.fam’ file (similar to
PLINK’s conventions).</p></li>
<li><p>Identify missing values in both samples and
SNPs/features.</p></li>
<li><p>Impute missing values per user’s specified method. See <a href="https://privefl.github.io/bigsnpr/reference/snp_fastImputeSimple.html" class="external-link">R
documentation for <code>bigsnpr::snp_fastImputeSimple()</code></a> for
more details. <strong>Note: the plmmr package cannot fit models if
datasets have missing values.</strong> All missing values must be
imputed or subset out before analysis.</p></li>
<li><p>Integrate in the external predictor information, if supplied.
This could be a matrix of meta-data (e.g., age, principal components,
etc.). <strong>Note:</strong> If there are samples in the supplied file
that are not included in the PLINK data, these will be removed. For
example, if you have more phenotyped participants than genotyped
participants in your study, <code><a href="../reference/process_plink.html">plmmr::process_plink()</a></code> will
create a matrix of data representing all the <em>genotyped</em> samples
that also have data in the supplied external phenotype file.</p></li>
<li><p>Create a design matrix that represents the nonsingular features
and the samples that have predictor and phenotype information available
(in the case where external data are supplied).</p></li>
<li><p>Standardize the design matrix so that all columns have mean of 0
and variance of 1.</p></li>
</ul></main>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Tabitha K. Peter, Anna C. Reisetter, Patrick J. Breheny, Yujing Lu.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

    </footer>
</div>





  </body>
</html>
